pyxrf/model/fit_spectrum.py

```diff
@@ -256,17 +274,222 @@ def energy_bound_low_update(self, change):
+        # The error: the original center position for the peak was used to
+        #   compute sigma of the Gaussian approximation of the peak. It almost not matter
+        #   for element peaks, since they don't move along the energy axis. But it is
+        #   invalid for Userpeaks, which are placed at a fixed position (5 keV) and then
+        #   moved to the desired position by changing ``delta_center``. Sigma depends
+        #   on actual center position, so ``delta_center`` should be taken into account.
+        #
+        # The original code for computing Gaussian function:
+        # def get_sigma(center):
+        #     temp_val = 2 * np.sqrt(2 * np.log(2))
+        #     return np.sqrt((fwhm_offset / temp_val) ** 2 + center * epsilon * fwhm_fanoprime)
+        #
+        # x = e_offset + x * e_linear + x ** 2 * e_quadratic
+        #
+        # return gaussian(x, area, center + delta_center,
+        #                 delta_sigma + get_sigma(center)) * ratio * ratio_adjust
+        #
+        # To correct the error, the last statement should be changed to the following:
+        # return gaussian(x, area, center + delta_center,
+        #                 delta_sigma + get_sigma(center + delta_center)) * ratio * ratio_adjust
+
+        # Temporary fix for the error mentioned above (to keep peak hight consistent as it moves
+        #   along the enery axis, computation of sigma must be identical here and in
+        #   the function ``element_peak_xrf``
+        sigma_sqr = 5.0  # center (ignore ``delta_sigma``)
+        # The next line is correct (but may be enabled only if the bug in scikit-beam is corrected)
+        # sigma_sqr = self.param_dict[self.name_userpeak_dcenter]["value"] + 5.0  # center
+
+        sigma_sqr *= self.param_model.default_parameters["non_fitting_values"]["epsilon"]  # epsilon
+        sigma_sqr *= self.param_model.default_parameters["fwhm_fanoprime"]["value"]  # fanoprime
+        sigma_sqr += sigma * sigma  # We have computed the expression under sqrt
+
+        sigma_total = np.sqrt(sigma_sqr)
+
+        return self._sigma_to_fwhm(sigma_total)
+
+    def _sigma_to_fwhm(self, sigma):
+        # Converts 'sigma' to 'fwhm'
+        v_const = 2 * np.sqrt(2 * np.log(2))
+        return sigma * v_const
+
+    def _fwhm_to_sigma(self, fwhm):
+        # Converts 'fwhm' to 'sigma'
+        v_const = 2 * np.sqrt(2 * np.log(2))
+        return fwhm / v_const
+
+    def _gaussian_max_to_area(self, peak_max, peak_sigma):
+        # Returns the area of Gaussian peak for the given maximum and sigma
+        s2pi = np.sqrt(2 * np.pi)
+        return peak_max * s2pi * peak_sigma
+
+    def _gaussian_area_to_max(self, peak_area, peak_sigma):
+        # Returns maximum of the area of Gaussian peak for the given peak area and sigma
+        s2pi = np.sqrt(2 * np.pi)
+        if peak_sigma == 0:
+            return 0
+        else:
+            return peak_area / s2pi / peak_sigma
```

Comment:
I wonder if we can use a library which does these pretty standard operations, instead of inventing ours? `peakutils` or something else should have an option.

We have some relevant code in https://github.com/bluesky/bluesky/blob/master/bluesky/callbacks/fitting.py, maybe they can be combined or isolated into scikit-beam utils?

Reply:
I don't think these are standard operations. They are specific for certain representation of Gaussian function and don't represent general interest. The only reason I had to use them here is the way the parameters are represented throughout the program. The transformation are also trivial. They may be moved to scikit-beam later. I moved them to ``utils.py`` for now.

I looked at ``fitting.py``. I am not sure that those functions are related to fitting functions in ``scikit-beam``, except that both are calling optimization procedures from ``lmfit``. But I think that making ``scikit-beam``, especially in its current state, a dependency for Bluesky is not a good idea.